### PR TITLE
[Doc] Add developer guide for CustomOp

### PR DESCRIPTION
**Pull request by shen-shanshan**
## Purpose

Currently, there are more and more `CustomOp` defined both in vLLM and other OOT plugin devices. Following https://github.com/vllm-project/vllm/pull/30125#issuecomment-3648916991, I have written a doc about the principle and usage of `CustomOp`.

## Test Result

---
<details>
<summary> Essential Elements of an Effective PR Description Checklist </summary>

- [x] The purpose of the PR, such as "Fix some issue (link existing issues this PR will resolve)".
- [ ] The test plan, such as providing test command.
- [ ] The test results, such as pasting the results comparison before and after, or e2e results
- [ ] (Optional) The necessary documentation update, such as updating `supported_models.md` and `examples` for a new model.
- [ ] (Optional) Release notes update. If your change is user facing, please update the release notes draft in the [Google Doc](https://docs.google.com/document/d/1YyVqrgX4gHTtrstbq8oWUImOyPCKSGnJ7xtTpmXzlRs/edit?tab=t.0).
</details>

---

> [!NOTE]
> Introduces a developer guide for `CustomOp` explaining registration, enable/disable semantics, backend dispatch, and OOT plugin integration, plus customization examples.
> 
> - Adds `docs/design/custom_op.md` with how-to, config controls, supported ops list, and OOT plugin registration examples
> - Inserts snippet anchors (`--8<-- [start:...]`/`[end:...]`) around numerous in-tree op classes (attention, activation, conv, embedding/LM head, linear, logits processor, Mamba, MoE, norms, quantization, RoPE) to enable docs inclusion
> - Minor docstring/comment clarifications (e.g., `CompilationConfig.custom_ops` default behavior wording; notes in `CustomOp.dispatch_forward`); no logic changes
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 27b46eb95f2f81b3953184da148367659aa620dd. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> <sup>[Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) is generating a summary for commit 42e8aa793297f6e260711b576a294f4b4ceeb381. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

---

> [!NOTE]
> Introduces a developer guide for `CustomOp` and prepares code for docs inclusion.
> 
> - Adds `docs/design/custom_op.md` covering registration, backend dispatch, enable/disable controls, OOT plugin integration, and examples
> - Inserts snippet anchors (`--8<-- [start:...]/[end:...]`) around numerous in-tree ops (attention, activations, conv, embedding/LM head, linear, logits processor, Mamba, MoE, norms, quantization, RoPE) to enable reuse in docs
> - Minor documentation/comment tweaks: refine `CompilationConfig.custom_ops` default description and add notes in `CustomOp.dispatch_forward`; no logic changes
> 
> <sup>Written by [Cursor Bugbot](https://cursor.com/dashboard?tab=bugbot) for commit 42e8aa793297f6e260711b576a294f4b4ceeb381. This will update automatically on new commits. Configure [here](https://cursor.com/dashboard?tab=bugbot).</sup>

### CODE REVIEW

File: docs/design/custom_op.md [hunk 1]
```diff
@@ -0,0 +1,236 @@
...
+For example:
+
+- Use `--compilation_config.custom_ops '["all"]'` to enable all custom ops.
+- Use `--compilation_config.custom_ops '["none"]'` to disable all custom ops.
+- Use `--compilation_config.custom_ops '["all,-op1"]'` to enable all custom ops except op1 (i.e., prefixed with a `-` means "disable").
+- Use `--compilation_config.custom_ops '["none,+op1,+op2"]'` to only enable op1 and op2 (i.e., prefixed with a `+` means "enable").
+
+## Types of Supported CustomOp in vLLM
+
+| Category | OP Name | OP Class |
```
Comment (ZJY0516) [inline docs/design/custom_op.md]: Do we really need to list all custom op here?

File: docs/design/custom_op.md [hunk 2]
```diff
@@ -0,0 +1,236 @@
...
+- **TPU platform:** dispatch to `forward_tpu()`.
+- **OOT platform:** dispatch to `forward_oot()`. This will only be called on OOT platforms.
+- **Default:** dispatch to `forward_native()` as a final fallback for all platforms.
+
+Furthur more, vLLM decides whether enable or disable a `CustomOp` by `compilation_config.custom_ops`. To be specific, if a `CustomOp` is not registered (i.e., use default config), it will be enabled if there is a `all` in `compilation_config.custom_ops` or will be disabled if there is a `none`.
+
+!!! note
+    Note that `all` and `none` cannot coexist in `compilation_config.custom_ops`.
+
+By default, if `compilation_config.backend == "inductor"` and `compilation_config.mode != CompilationMode.NONE`, a `none` will be appended into `compilation_config.custom_ops`, otherwise a `all` will be appended. In other words, this means `CustomOp` will be disabled in some platforms (i.e., those use `inductor` as dafault backend for `torch.compile`) when running with graph mode. In this case, Inductor generates (fused) Triton kernels for those disabled custom ops.
```
Comment (ZJY0516) [inline docs/design/custom_op.md]: Perhaps we should clarify what "graph mode" means in this context.

File: docs/design/custom_op.md (hunk 1 above)
Comment (tjtanaa) [reply docs/design/custom_op.md]: I think it is a good start to know how many custom ops are there as they are all scattered around the code and there is no cli command or anyway to list them all out.

File: docs/design/custom_op.md (hunk 2 above)
Comment (shen-shanshan) [reply docs/design/custom_op.md]: > Perhaps we should clarify what "graph mode" means in this context.

"graph mode" means `compilation_config.mode > CompilationMode.NONE`, find more details at https://github.com/vllm-project/vllm/blob/b10d47e0e099b29aecc20740daaaa1afe3476c59/vllm/config/compilation.py#L36-L49.

It indicates that inference with `torch.compile`, not "eager mode".

File: docs/design/custom_op.md (hunk 1 above)
Comment (hmellor) [reply docs/design/custom_op.md]: Perhaps it is better to use a [snippet](https://facelessuser.github.io/pymdown-extensions/extensions/snippets/#snippet-sections) from the code where it is defined so that this doesn't go stale immediately?

File: docs/design/custom_op.md (hunk 2 above)
Comment (hmellor) [reply docs/design/custom_op.md]: In that case let's call it "torch compile mode" so that it's not confused with CUDA Graphs mode

File: vllm/config/compilation.py [hunk 3]
```diff
@@ -404,7 +404,8 @@ class CompilationConfig:
     - 'none,+op1,+op2' to enable only op1 and op2
 
     By default, all custom ops are enabled when running without Inductor and
-    disabled when running with Inductor: mode>=VLLM_COMPILE and backend="inductor".
+    disabled when running with Inductor: mode>=CompilationMode.NONE and
```
Comment (hmellor) [inline vllm/config/compilation.py]: @ProExpertProg is this correct?

File: docs/design/custom_op.md (hunk 2 above)
Comment (ZJY0516) [reply docs/design/custom_op.md]: > In that case let's call it "torch compile mode" so that it's not confused with CUDA Graphs mode

Agreed

File: vllm/config/compilation.py (hunk 3 above)
Comment (ProExpertProg) [reply vllm/config/compilation.py]: I didn't think so, but if it is, it's wrong: we should always disable custom ops by default if inductor is used (for any compilation mode > 0)

Comment (hmellor) [reply vllm/config/compilation.py]: So the change in this PR is correct?

Comment (ProExpertProg) [reply vllm/config/compilation.py]: ~No the change in this PR is incorrect~

EDIT: sorry I misread this. The change is correct apart from `mode>CompilationMode.NONE` instead of `mode>=CompilationMode.NONE`

File: docs/design/custom_op.md (hunk 2 above)
Comment (shen-shanshan) [reply docs/design/custom_op.md]: ok.

File: vllm/config/compilation.py (hunk 3 above)
Comment (shen-shanshan) [reply vllm/config/compilation.py]: > ~No the change in this PR is incorrect~
> 
> EDIT: sorry I misread this. The change is correct apart from `mode>CompilationMode.NONE` instead of `mode>=CompilationMode.NONE`

ok.

File: docs/design/custom_op.md (hunk 1 above)
Comment (shen-shanshan) [reply docs/design/custom_op.md]: > Perhaps it is better to use a [snippet](https://facelessuser.github.io/pymdown-extensions/extensions/snippets/#snippet-sections) from the code where it is defined so that this doesn't go stale immediately?

Let me take a look.

Comment (shen-shanshan) [reply docs/design/custom_op.md]: @hmellor I have updated the CustomOp list using [snippet](https://facelessuser.github.io/pymdown-extensions/extensions/snippets/#snippet-sections). Could you please take a look at this about whether it's appropriate?

File: docs/design/custom_op.md [hunk 4]
```diff
@@ -0,0 +1,39 @@
...
+### Implement a new CustomOp in vLLM
+
+...
+
+### Implement a new CustomOp in OOT device plugins
+
+Currently, thanks to [vLLM's hardware-plugin mechanism](./plugin_system.md), there are various OOT (Out-Of-Tree) device plugins emerge out to enable vLLM seamlessly runs on different hardwares. You can also find more details about this mechanism at [Introducing vLLM Hardware Plugin, Best Practice from Ascend NPU](https://blog.vllm.ai/2025/05/12/hardware-plugin.html).
+
+- **Official device plugins:** [vllm-ascend](https://github.com/vllm-project/vllm-ascend) (for Huawei Ascend NPU), [vllm-spyre](https://github.com/vllm-project/vllm-spyre)
+(for Spyre), [vllm-gaudi](https://github.com/vllm-project/vllm-gaudi) (for Intel Gaudi), [vllm-neuron](https://github.com/vllm-project/vllm-neuron) (for AWS Neuron), [vllm-meta](https://github.com/vllm-project/vllm-metal) (for Apple Silicon), etc.
```
[the remark this answers is not in the document]
Comment (ILikeIneine) [reply docs/design/custom_op.md]: seems a typo here
meta -> metal

**Comment by shen-shanshan** [comment]
@hmellor @tjtanaa @ProExpertProg Any other suggestions about this PR? Could we merge this? It has been hanging for a long time. 😄

File: docs/design/custom_op.md [hunk 5]
```diff
@@ -0,0 +1,318 @@
...
+    ```python
+    class CustomOp(nn.Module):
+
+        op_registry: dict[str, type["CustomOp"]] = {}
+        op_registry_oot: dict[str, type["CustomOp"]] = {}
+    ```
+
+We can use `@CustomOp.register("op_name")` to register an op class to the `CustomOp` system. After this, the `op_name` and its class will be added into the `op_registry` dictionary. In addition, We can also register an OOT op by `@CustomOp.register_oot("op_name")`. We will introduce this mechanism in detail later.
+
+When a `CustomOp` is called (i.e., call its `forward()` method), if it is enabled, it will automatically dispatch the forward method to the appropriate backend according to `current_platform`. Otherwise (i.e., it is disabled), it will only call the `forward_native()` method to use PyTorch-native implementation of this forward method.
```
Comment (tjtanaa) [inline docs/design/custom_op.md]: @shen-shanshan I hope it is not late. I have just reread again. Can we make it clearer what does it mean by enabled here?
like


When a `CustomOp` is called (i.e., call its `forward()` method), and it is enabled i.e. `--compilation_config.custom_ops '["+op_name"]'`, it will automatically dispatch the forward method to the appropriate backend according to `current_platform`.

File: docs/design/custom_op.md [hunk 6]
```diff
@@ -0,0 +1,318 @@
...
+- **ROCm platform:** dispatch to `forward_hip()`. If `forward_hip()` is not implemented, it will use `forward_cuda()` as a fallback.
+- **XPU platform:** dispatch to `forward_xpu()`.
+- **TPU platform:** dispatch to `forward_tpu()`.
+- **OOT platform:** dispatch to `forward_oot()`. This will only be called on OOT platforms.
+- **Default:** dispatch to `forward_native()` as a final fallback for all platforms.
+
+!!! note
+    Note that the dispatching logic might not be absolute because of class inheritance. Derived class might override the behavior.
+
+Furthur more, vLLM decides whether enable or disable a `CustomOp` by `compilation_config.custom_ops`. To be specific, if a `CustomOp` is not registered (i.e., use default config), it will be enabled if there is a `all` in `compilation_config.custom_ops` or will be disabled if there is a `none`.
```
Comment (tjtanaa) [inline docs/design/custom_op.md]: NITs: 

Furthermore, vLLM decides whether to enable or disable a CustomOp based on `compilation_config.custom_ops`. To be specific, if a CustomOp is not registered in `compilation_config.custom_ops` (i.e., uses the default config), it will be enabled if `compilation_config.custom_ops` contains `all`, or will be disabled if it contains `none`.

**Review by tjtanaa — APPROVED** [review]
LGTM, just some more changes to make the important points clearer and explicit.

File: docs/design/custom_op.md (hunk 1 above)
Comment (hmellor) [reply docs/design/custom_op.md]: I think this still suffers from the issue that it will go stale once more custom ops are added.

The rendered doc does look very nice though. I wonder if we could achieve the same thing with `mkdocstrings` references so we don't have to manually annotate all the custom ops in the source.

Could you make a follow up PR that tries this? It'd look something like

```md
::: vllm.model_executor.layers.activation.SiluAndMul
```

Comment (shen-shanshan) [reply docs/design/custom_op.md]: > I think this still suffers from the issue that it will go stale once more custom ops are added.
> 
> The rendered doc does look very nice though. I wonder if we could achieve the same thing with `mkdocstrings` references so we don't have to manually annotate all the custom ops in the source.
> 
> Could you make a follow up PR that tries this? It'd look something like
> 
> ```
> ::: vllm.model_executor.layers.activation.SiluAndMul
> ```

Thanks for your suggestion, I will try it out when I have free time.
